### ISSUE TITLE
Post Media: Hashing

### ISSUE THREAD

**Issue by abitgone** [appdotnet/api-spec#80]
To allow for better caching of media when included in a post, I'd like to propose that each media object contains a hash of the media object that has been added:

``` json
{
    "entities": {
        "media": [{
            "height": 512,
            "width": 512,
            "mimetype": "image/jpeg",
            "url": "https://example.com/image.jpg",
            "hash": "m29gCOjAMr+HefJXo0ujnAlIO/I91Rj1aE9ablq1VI9ycgXfFB4tqFyf42fzG8CoIuiiXq2SKrzXbsrjR2sqdg=="
        }]
    }
}
```

This would allow clients, particularly on bandwidth-constrained connections, to selectively download and cache content based on the hash, as opposed to downloading the same file again and again from different locations when the people you follow have decided to re-post the same media object a dozen times.

Admittedly, this is somewhat of a rage-induced feature request – since I see the same photo posted again, and again, and again on twitter and would dearly love the ability to block the same image after I've seen it 27 times (having the ability to tell my client not to bother downloading posts that contain the same media object I'm tired of seeing again and again) – though I'm sure the caching idea is a much better use-case.

**Comment by tgvashworth** [comment]
That's actually a really good idea. +1.

**Comment by michaelrhanson** [comment]
If this is implemented, let's be explicit about the use of this hash as the ETag in the HTTP response, so HTTP caching schemes can work correctly with it.

**Comment by abitgone** [comment]
@michaelrhanson Interesting idea, but how can you influence the ETag that's returned – potentially – from multiple different image hosting services?

**Comment by abitgone** [comment]
@michaelrhanson Additionally, can you absolutely rock-solidly rely on the same ETag being served from different web servers serving an identical image, audio or video file?

**Comment by michaelrhanson** [comment]
Hm.  No, you can't.  That's a problem.

But it also seems like introducing this feature - as an application-level re-implementation of the ETag idea - could cause some pretty twisty code for clients.  The URL is already supposed to be a unique identifier of the resource; if the client needs to maintain a Last-Modified, ETag, and AppNet-Hash for each of these, and properly handle both application and network-level caching... could be tricky. 

I'll try turning my suggestion into a question: what use cases can this support that HTTP caching doesn't handle already? (and mea culpa on http://infrequently.org/2012/08/inadmissible-arguments/ )

**Comment by abitgone** [comment]
HTTP caching wouldn't prevent clients from downloading an identical image  stored on multiple different websites. Identifying whether a media object has already been cached locally (or seen/downloaded/cached locally already) via a separate hash would be trivial using this method, but impossible using Etags or HTTP caching.

**Comment by jschlesser** [comment]
I think that media should be able to be handled from a variety of vendors, all the CDNs, amazon s3 etc...
Clients can obey cache control directives already, no need to reinvent.

Please clarify the initial cause of rage.

Same photo with different URLs causing the same photo to be downloaded?
Same photo with same URL but poor HTTP cache control?

**Comment by abitgone** [comment]
This is more a case of dealing with an identical file that has been shared on multiple different websites.

The initial cause of this was an image that has been downloaded from a source and posted on different photo sharing websites, and re-posted again by many people who then get others to re-post the same image. So let's go from there. We have our image. It is posted to ADN, via an image sharing network, and it becomes quite popular. 

One of the following things will happen to the post containing the image:
- **Native Repost**
  Because of it's popularity, many ADN users use the native repost feature, which means an identical copy of my original post is sent out to their stream. Same file, same URL endpoint, so no problems there – HTTP caching will do its job here.
- **Copy-and-Repost**
  Some ADN users will choose to simply copy the body text of the original post, link to image included, which means an identical copy of the image URL is sent out. Again, this is the same file and the same URL endpoint, so no problems here either – HTTP caching again does its job.
- **Download Image and Post As Own**
  Other ADN users may choose to download the file and repost it themselves. The image will, in all likelihood, be sent to another image sharing network. The image is identical to the one that has already been uploaded. HTTP caching can't catch this case as the URL endpoint is likely different. 
- **Multiple people grabbing the same source file from [popular website] and posting it**
  Chances are many people will upload the same file to many different networks. In cases where those users have a common follower, that particular user will benefit from being able to know they've already seen that image, or that they already have a copy of it. 

Additionally, if the API allowed clients to submit the hash of the media object it was uploading, it could be smart enough to say "hey, you don't need to upload that to [your service] – use this URL instead", saving clients from having to re-upload the media object over and over.

**Comment by ralfr** [comment]
+1 Would be a real good differentiator, completely optional for client devs and can speed up performance significantly.

**Comment by tgvashworth** [comment]
Yeah I agree this is neat because it's totally optional. Although, does it really need to be part of the spec? Is it possible, or should it be possible, to attach arbitrary meta-data to an entity as it is to a post?
